src/engine/client/audio.cpp

```diff
@@ -0,0 +1,330 @@
+/* Audio Source */
+
+AudioSource::AudioSource(unsigned int al_sample, float gain, float pitch, bool looping) :
+        AudioSource(al_sample, PLVector3(0, 0, 0), PLVector3(0, 0, 0), gain, pitch, looping) {
+    alSourcei(al_source_id_, AL_SOURCE_RELATIVE, 1);
+}
+
+AudioSource::AudioSource(unsigned int al_sample, PLVector3 pos, PLVector3 vel, float gain, float pitch, bool looping) {
+    alGenSources(1, &al_source_id_);
+    OALCheckErrors();
+
+    SetPosition(pos);
+    SetVelocity(vel);
+    SetGain(gain);
+    SetPitch(pitch);
+
+    alSourcei(al_source_id_, AL_LOOPING, looping);
+    OALCheckErrors();
+
+    alSourceQueueBuffers(al_source_id_, 1, &al_sample);
+
+    AudioManager::GetInstance()->sources_.insert(this);
+}
+
+AudioSource::~AudioSource() {
+    AudioManager::GetInstance()->sources_.erase(this);
+
+    alDeleteSources(1, &al_source_id_);
+}
+
+void AudioSource::SetPosition(PLVector3 position) {
+    alSource3f(al_source_id_, AL_POSITION, position.x, position.y, position.z);
+    OALCheckErrors();
+    position_ = position;
+}
+
+void AudioSource::SetVelocity(PLVector3 velocity) {
+    alSource3f(al_source_id_, AL_VELOCITY, velocity.x, velocity.y, velocity.z);
+    OALCheckErrors();
+    velocity_ = velocity;
+}
+
+void AudioSource::SetGain(float gain) {
+    alSourcef(al_source_id_, AL_GAIN, gain);
+    OALCheckErrors();
+    gain_ = gain;
+}
+
+void AudioSource::SetPitch(float pitch) {
+    alSourcef(al_source_id_, AL_PITCH, pitch);
+    OALCheckErrors();
+    pitch_ = pitch;
+}
+
+void AudioSource::StartPlaying() {
+    alSourcePlay(al_source_id_);
+    OALCheckErrors();
+}
+
+void AudioSource::StopPlaying() {
+    int state;
+    alGetSourcei(al_source_id_, AL_SOURCE_STATE, &state);
+    OALCheckErrors();
+    if(state != AL_PLAYING) {
+        return;
+    }
+
+    alSourceStop(al_source_id_);
+}
+
+/************************************************************/
+
+LPALGENEFFECTS alGenEffects;
+LPALDELETEEFFECTS alDeleteEffects;
+LPALISEFFECT alIsEffect;
+LPALEFFECTI alEffecti;
+LPALEFFECTIV alEffectiv;
+LPALEFFECTF alEffectf;
+LPALEFFECTFV alEffectfv;
+LPALGETEFFECTI alGetEffecti;
+LPALGETEFFECTIV alGetEffectiv;
+LPALGETEFFECTF alGetEffectf;
+LPALGETEFFECTFV alGetEffectfv;
+
+LPALGENAUXILIARYEFFECTSLOTS alGenAuxiliaryEffectSlots;
+LPALDELETEAUXILIARYEFFECTSLOTS alDeleteAuxiliaryEffectSlots;
+LPALISAUXILIARYEFFECTSLOT alIsAuxiliaryEffectSlot;
+LPALAUXILIARYEFFECTSLOTI alAuxiliaryEffectSloti;
+LPALAUXILIARYEFFECTSLOTIV alAuxiliaryEffectSlotiv;
+LPALAUXILIARYEFFECTSLOTF alAuxiliaryEffectSlotf;
+LPALAUXILIARYEFFECTSLOTFV alAuxiliaryEffectSlotfv;
+LPALGETAUXILIARYEFFECTSLOTI alGetAuxiliaryEffectSloti;
+LPALGETAUXILIARYEFFECTSLOTIV alGetAuxiliaryEffectSlotiv;
+LPALGETAUXILIARYEFFECTSLOTF alGetAuxiliaryEffectSlotf;
+LPALGETAUXILIARYEFFECTSLOTFV alGetAuxiliaryEffectSlotfv;
+
+AudioManager *AudioManager::instance_ = nullptr;
+
+AudioManager::AudioManager() {
+    memset(al_extensions_, false, MAX_AUDIO_EXT_SLOTS);
```

Comment:
I would recommend using a loop to set these to false. Or change false to zero and the size to sizeof(al_extensions_)